onnx/defs/controlflow/defs.cc

```diff
@@ -289,6 +289,16 @@ void LoopInferenceFunction(InferenceContext& ctx) {
       if (seq_type.has_elem_type() && seq_type.elem_type().has_tensor_type()) {
         seq_type.mutable_elem_type()->mutable_tensor_type()->clear_shape();
       }
+    } else if (input_type.has_optional_type()) {
```

Comment:
It may be helpful to extract lines 285 to 301 or so as a utility function ```ClearShape(TypeProto&)```. If the function is made recursive, it can handle arbitrary types a bit more easily. (Will help in the future if we expand the types even more).